storage/src/storage_backend.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) SimpleStaking and Tezedge Contributors
+// SPDX-License-Identifier: MIT
+
+use crate::persistent::database::{DBError, RocksDBStats};
+use failure::Fail;
+use serde::Serialize;
+use std::collections::HashSet;
+use std::mem;
+
+use crate::merkle_storage::{ContextValue, EntryHash};
+
+pub fn size_of_vec<T>(v: &Vec<T>) -> usize {
+    mem::size_of::<Vec<T>>() + mem::size_of::<T>() * v.capacity()
+}
+
+#[derive(Debug, Fail)]
+pub enum StorageBackendError {
+    #[fail(display = "RocksDB error: {}", error)]
+    RocksDBError { error: rocksdb::Error },
+    #[fail(display = "Column family {} is missing", name)]
+    MissingColumnFamily { name: &'static str },
+    #[fail(display = "Backend Error")]
+    BackendError,
+    #[fail(display = "SledDB error: {}", error)]
+    SledDBError { error: sled::Error },
+    #[fail(display = "Guard Poison {} ", error)]
+    GuardPoison { error: String },
+    #[fail(display = "Serialization error: {:?}", error)]
+    SerializationError { error: bincode::Error },
+    #[fail(display = "DBError error: {:?}", error)]
+    DBError { error: DBError },
+}
+
+impl From<rocksdb::Error> for StorageBackendError {
+    fn from(error: rocksdb::Error) -> Self {
+        StorageBackendError::RocksDBError { error }
+    }
+}
+
+impl From<sled::Error> for StorageBackendError {
+    fn from(error: sled::Error) -> Self {
+        StorageBackendError::SledDBError { error }
+    }
+}
+
+impl From<DBError> for StorageBackendError {
+    fn from(error: DBError) -> Self {
+        StorageBackendError::DBError { error }
+    }
+}
+
+impl From<bincode::Error> for StorageBackendError {
+    fn from(error: bincode::Error) -> Self {
+        StorageBackendError::SerializationError { error }
+    }
+}
+
+impl slog::Value for StorageBackendError {
+    fn serialize(
+        &self,
+        _record: &slog::Record,
+        key: slog::Key,
+        serializer: &mut dyn slog::Serializer,
+    ) -> slog::Result {
+        serializer.emit_arguments(key, &format_args!("{}", self))
+    }
+}
+
+pub trait StorageBackend: Send + Sync {
+    fn is_persisted(&self) -> bool;
+    fn get(&self, key: &EntryHash) -> Result<Option<ContextValue>, StorageBackendError>;
+    fn put(&mut self, key: &EntryHash, value: ContextValue) -> Result<bool, StorageBackendError>;
+    fn merge(&mut self, key: &EntryHash, value: ContextValue) -> Result<(), StorageBackendError>;
+    fn delete(&mut self, key: &EntryHash) -> Result<Option<ContextValue>, StorageBackendError>;
+    fn contains(&self, key: &EntryHash) -> Result<bool, StorageBackendError>;
+    fn get_mem_use_stats(&self) -> Result<RocksDBStats, StorageBackendError>;
```

Comment:
It is not always `RocksDBStats`, right? I saw that the `in_memory_storage` backend updates a different kind of stats, but always returns an empty `RocksDBStats` here

Reply:
yes - the stats part is  its coupled with GCs(then return type is extended). I just checked and it looks like for some of the backends `get_stats` was not implemented on the feature branch.

Reply:
@mat13mn sounds fine to me, and I guess those stats don't make as much sense as when the GC is involved. @bkontur  is that ok with you?

Reply:
@tizoc @mat13mn 
ok, just add here `TODO: TE-XYZ <some-desc-what-should-be-done>` please,
we can do that later

Reply:
done